module/zfs/vdev_indirect.c

```diff
@@ -889,44 +893,56 @@ spa_start_indirect_condensing_thread(spa_t *spa)
 }
 
 /*
- * Gets the obsolete spacemap object from the vdev's ZAP.
- * Returns the spacemap object, or 0 if it wasn't in the ZAP or the ZAP doesn't
- * exist yet.
+ * Gets the obsolete spacemap object from the vdev's ZAP.  On success sm_obj
+ * will contain either the obsolete spacemap object or DMU_NO_OBJECT if none
+ * exists.  All other errors are returned to the caller.
  */
 int
-vdev_obsolete_sm_object(vdev_t *vd)
+vdev_obsolete_sm_object(vdev_t *vd, uint64_t *sm_obj)
 {
 	ASSERT0(spa_config_held(vd->vdev_spa, SCL_ALL, RW_WRITER));
+
 	if (vd->vdev_top_zap == 0) {
+		*sm_obj = DMU_NO_OBJECT;
 		return (0);
 	}
 
-	uint64_t sm_obj = 0;
-	int err;
-	err = zap_lookup(vd->vdev_spa->spa_meta_objset, vd->vdev_top_zap,
-	    VDEV_TOP_ZAP_INDIRECT_OBSOLETE_SM, sizeof (sm_obj), 1, &sm_obj);
-
-	ASSERT(err == 0 || err == ENOENT);
+	int error = zap_lookup(vd->vdev_spa->spa_meta_objset, vd->vdev_top_zap,
+	    VDEV_TOP_ZAP_INDIRECT_OBSOLETE_SM, sizeof (sm_obj), 1, sm_obj);
+	if (error == ENOENT) {
+		*sm_obj = DMU_NO_OBJECT;
+		error = 0;
+	}
 
-	return (sm_obj);
+	return (error);
 }
 
-boolean_t
-vdev_obsolete_counts_are_precise(vdev_t *vd)
+/*
+ * Gets the obsolete count are precise spacemap object from the vdev's ZAP.
+ * On success are_precise will be set to reflect is the counts are precise.
+ * All other errors are returned to the caller.
```

Comment:
... which all callers assert/verify is 0.  So there's no behavior change but I guess this is ready for callers that can handle the error?

Reply:
Right, that was my intent.  This way the import path could be updated and the other callers could functionally remain unchanged.  None of the other callers were set up to be able to do anything reasonable in case of an error, and I wanted to keep the change small.